dht.go

```diff
@@ -372,11 +403,26 @@ func (dht *IpfsDHT) putLocal(key string, rec *recpb.Record) error {
 	return dht.datastore.Put(mkDsKey(key), data)
 }
 
-// Update signals the routingTable to Update its last-seen status
-// on the given peer.
-func (dht *IpfsDHT) Update(ctx context.Context, p peer.ID) {
-	logger.Event(ctx, "updatePeer", p)
-	dht.routingTable.Update(p)
+// peerFound signals the routingTable that we've found a peer that
+// supports the DHT protocol.
+func (dht *IpfsDHT) peerFound(ctx context.Context, p peer.ID) {
```

Comment:
@aarshkshah1992 and I discussed this a bit offline. As long as the routing table has a bounded queue size for connected peers it is possible that a peer we are connected to will not be added to either the routing table or the replacement queue. If that occurs then we have no event (other than `fixLowPeers` if the entire routing table is almost completely empty) that cause us to re-add existing peers to the routing table/replacement queue.

Some fixes for this include:
1) Maintaining state: e.g. allowing the replacement cache to contain an unbounded number of connected peers
2) Polling state: loop through our connected peers more frequently (e.g. periodically, when a bucket goes from full to nothing in the replacement cache, etc.)

Reply:
I don't like the idea of an unbounded replacement cache as it will cause leaks for Cpls that don't see missing peers often.

**2** sounds like a good idea. We can add a connected peer that is not currently in the replacement cache AND RT to the cache when we drop a peer from the replacement cache i..e make it event driven. The DHT could pass in a callback to the RT to get a candidate for this purpose.


WDYT @Stebalien ? Should we do this right away ?

Reply:
I'd do this but punt it to a future patch. Could you file an issue?

My only concern is that we don't want to _frequently_ loop over the entire list of connections if we happen to not know of any DHT server peers.

Reply:
Here's the issue:

https://github.com/libp2p/go-libp2p-kbucket/issues/58